### PR TITLE
gh-117534: Add checking for input parameter in iso_to_ymd

### PR DESCRIPTION
**Pull request by Vlad4896**
Python terminates with core dumped on the following test:
```
import datetime

y = datetime.datetime.fromisoformat('0000W25')
print(y)
```
The test result:
```
python: ./Modules/_datetimemodule.c:290: days_before_year: Assertion `year >= 1' failed.
Aborted (core dumped)
```
The issue happened due to lack of checking for input parameters in iso_to_ymd().

* Issue: gh-117534

### CODE REVIEW

File: Modules/_datetimemodule.c
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: -2 means a specific kind of error (wrong week).

I think that the year value should be checked before calling `iso_to_ymd()`. It is already checked in one case, but not in other.

Please add tests.

Comment (Vlad4896) [reply Modules/_datetimemodule.c]: Based on the back trace, the year value is taken from the string at parse_iso format_date() (the caller of the iso_to_umd()). Do you recommend check it in parse_iso format_date() and return -1 (-1:  Failed to parse date component)? 
```
#10 0x00007ffff7e3d7d6 in iso_to_ymd (iso_year=0, iso_week=25, iso_day=1, year=year@entry=0x7fffffffd71c, month=month@entry=0x7fffffffd720, day=day@entry=0x7fffffffd724)
    at ./Modules/_datetimemodule.c:440
#11 0x00007ffff7e3d998 in parse_isoformat_date (dtstr=dtstr@entry=0x7ffff74dd328 "0000W25", len=len@entry=7, year=year@entry=0x7fffffffd71c, month=month@entry=0x7fffffffd720, 
    day=day@entry=0x7fffffffd724) at ./Modules/_datetimemodule.c:803
#12 0x00007ffff7e479c0 in datetime_fromisoformat (cls=0x7ffff7e58000 <PyDateTime_DateTimeType>, dtstr=0x7ffff74dd300) at ./Modules/_datetimemodule.c:5506
```
Do you mean add testcase to 'def test_fromisoformat_date_examples(self)' ?

Comment (serhiy-storchaka) [reply Modules/_datetimemodule.c]: Currently, there are two calls of `iso_to_umd()`. In one case the year value is validated before calling `iso_to_umd()`, in other case it is not validated. We should either move the validation into `iso_to_umd()` (and introduce a special error code for this), or always validate the year value before calling `iso_to_umd()`.

Comment (Vlad4896) [reply Modules/_datetimemodule.c]: Thank you for comments. The fix is updated accordingly. 
Also I added the test case to Lib/test/datetimetester.py.

**Comment by pganssle** [comment]
Thanks for this, and good noticing this edge case!

What you have so far looks good to me, other than the NEWS entry.

I haven't been able to test it out, but does this also work in the Pure Python version? Looks like that version may already have [logic to deal with this](https://github.com/python/cpython/blob/4d4a6f1b6aea6dae131ac116f1735a38c3e32cd1/Lib/_pydatetime.py#L471-L472), whereas the equivalent logic in `_datetimemodule.c` is in [`date_fromisocalendar`](https://github.com/python/cpython/blob/4d4a6f1b6aea6dae131ac116f1735a38c3e32cd1/Modules/_datetimemodule.c#L3146-L3149).

File: Lib/test/datetimetester.py
Comment (serhiy-storchaka) [inline Lib/test/datetimetester.py]: Should not it be `'0000-W25-1'` for consistency with other examples?

Comment (Vlad4896) [reply Lib/test/datetimetester.py]: As far as I understand the separator '-' is optional symbol. So I added the ISO data string from my test script.

**Comment by Vlad4896** [comment]
> Thanks for this, and good noticing this edge case!
> 
> What you have so far looks good to me, other than the NEWS entry.
> 
> I haven't been able to test it out, but does this also work in the pure Python version? Looks like that version may already have [logic to deal with this](https://github.com/python/cpython/blob/4d4a6f1b6aea6dae131ac116f1735a38c3e32cd1/Lib/_pydatetime.py#L471-L472), whereas the equivalent logic in `_datetimemodule.c` is in [`date_fromisocalendar`](https://github.com/python/cpython/blob/4d4a6f1b6aea6dae131ac116f1735a38c3e32cd1/Modules/_datetimemodule.c#L3146-L3149).

I've added NEWS entry. 
Could you please clarify, what do you mean "work in the pure Python version"?

File: Lib/test/datetimetester.py
Comment (pganssle) [inline Lib/test/datetimetester.py]: Yes, it is optional, though I do think it would be better to use the form `'0000-W25-1'` , because it will be more obvious that it's failing because of the invalid value rather than because of some issue with the format (I think that's why I chose a clearly valid format with well-separated values for the tests in the first place).

I know it's a bit of a scope creep, but while we are here, it would probably be a good idea to round out the coverage to cover invalid day of week as well.

Here is a suggested patch that adds the extra test cases, and moves the surrogate test case to the end, so that all the "invalid component" tests are grouped together.

```suggestion
            '0000-W25-1',       # Invalid year
            '10000-W25-1',      # Invalid year
            '2020-W25-0',       # Invalid day-of-week
            '2020-W25-8',       # Invalid day-of-week
            '2009\ud80002\ud80028',     # Separators are surrogate codepoints
```

File: Misc/NEWS.d/next/C API/2024-04-08-09-44-29.gh-issue-117534.54ZE_n.rst
Comment (pganssle) [inline Misc/NEWS.d/next/C API/2024-04-08-09-44-29.gh-issue-117534.54ZE_n.rst]: Generally speaking we try to document the public-facing behavior of a patch, not the implementation details. Here's how I would word this:

```suggestion
Improve validation logic in the C implementation of :meth:`datetime.datetime.fromisoformat` to better handle invalid years. 
```

It is also customary to add "Patch by <Your name>" at the end of the news fragment, but that is not required, and I don't know your name, so I cannot include it in the suggestion.

File: Lib/test/datetimetester.py
Comment (Vlad4896) [reply Lib/test/datetimetester.py]: I've updated the test cases as suggested.

**Review by pganssle — CHANGES_REQUESTED** [review]
> Could you please clarify, what do you mean "work in the pure Python version"?

`datetime` is a [PEP 399](https://peps.python.org/pep-0399/) module, meaning there are two implementations, [`_pydatetime`](https://github.com/python/cpython/blob/24a2bd048115efae799b0a9c5dd9fbb7a0806978/Lib/_pydatetime.py) (which is written entirely in Python, and [`_datetimemodule.c`](https://github.com/python/cpython/blob/24a2bd048115efae799b0a9c5dd9fbb7a0806978/Modules/_datetimemodule.c), which is written in C. Most changes need to update both versions, because the logic is supposed to be identical.

In this case, I think it's OK, because the Python version already does the right thing because of the way the code is organized. That said, I tried to run this today and it looks like the core dump only happens if the code is compiled with assertions on — which by default it is not — so I think the severity of this bug is not as bad as I had feared.

Overall, I think this is a good patch. It brings the C code more in line with the Python code by unifying the error handling for invalid years, and prevents us from relying on an accidental feature of the way this was being parsed. I made two suggestions in the comments in-line, and if you want you can add yourself to [the ACKS file](https://github.com/python/cpython/blob/main/Misc/ACKS), but other than that, if the CI passes after the changes, this is looking good to merge, assuming @serhiy-storchaka doesn't have any objections.

File: Misc/ACKS
Comment (pganssle) [inline Misc/ACKS]: I think they want to keep this roughly in alphabetical order by last name.

**Review by pganssle — APPROVED** [review]
LGTM, other than moving your name around in the `ACKS` list.

@serhiy-storchaka Any other comments, or should we merge?

File: Misc/NEWS.d/next/C API/2024-04-08-09-44-29.gh-issue-117534.54ZE_n.rst
Comment (Vlad4896) [reply Misc/NEWS.d/next/C API/2024-04-08-09-44-29.gh-issue-117534.54ZE_n.rst]: Thank you for the comments. The text is updated.

File: Misc/ACKS
Comment (Vlad4896) [reply Misc/ACKS]: I'm sorry. Did not pay attention on comment.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Pull request merged**
